src/macaron/util.py

```diff
@@ -272,6 +274,104 @@ def send_post_http_raw(
     return response
 
 
+class StreamWriteDownloader:
+    """A class to handle writing a streamed download to a file."""
+
+    def __init__(self, file: BinaryIO) -> None:
+        """Initialise the class with the file path."""
+        self.file = file
+
+    def chunk_function(self, chunk: bytes) -> None:
+        """Write the chunk to the file."""
+        self.file.write(chunk)
+
+
+def download_file_with_size_limit(
+    url: str, headers: dict, file_path: str, timeout: int = 40, size_limit: int = 0
+) -> bool:
+    """Download a file with a size limit that will abort the operation if exceeded.
+
+    Parameters
+    ----------
+    url: str
+        The target of the request.
+    headers: dict
+        The headers to use in the request.
+    file_path: str
+        The path to download the file to.
+    timeout: int
+        The timeout in seconds for the request.
+    size_limit: int
+        The size limit in bytes of the downloaded file.
+        A download will terminate if it reaches beyond this amount.
+
+    Returns
+    -------
+    bool
+        True if the operation succeeded, False otherwise.
+    """
+    try:
+        with open(file_path, "wb") as file:
+            downloader = StreamWriteDownloader(file)
+            return stream_file_with_size_limit(url, headers, downloader.chunk_function, timeout, size_limit)
+    except OSError as error:
+        logger.error(error)
+        return False
+
+
+def stream_file_with_size_limit(
+    url: str, headers: dict, chunk_function: Callable[[bytes], None], timeout: int = 40, size_limit: int = 0
+) -> bool:
+    """Stream a file download and perform the passed function on the chunks of its data.
+
+    If data in excess of the size limit is received, this operation will be aborted.
+
+    Parameters
+    ----------
+    url: str
+        The target of the request.
+    headers: dict
+        The headers to use in the request.
+    chunk_function: Callable[[bytes], None]
+        The function to use with each downloaded chunk.
+    timeout: int
+        The timeout in seconds for the request.
+    size_limit: int
```

Comment:
At the moment, `size_limit` is an optional parameter with the defaults value of 0. Does this mean that you should never use `stream_file_with_size_limit` without providing a value for `size_limit` into the function call ?